### PR TITLE
pre-norm TransformerBlock bug fix

### LINKED ISSUE #3: pre-norm bug in transformer block
Pre-norm should be (as in [mingpt](https://github.com/karpathy/minGPT/blob/3ed14b2cec0dfdad3f4b2831f2b4a86d11aef150/mingpt/model.py#L97) code):
`x + attn(norm(x))`

In my code now:
```
norm_x = norm(x)
norm_x + attn(norm_x)
```
https://github.com/Howuhh/faster-trajectory-transformer/blob/cc1af3a7dfcd8ecc6d0ec0f719b7673329ec0c8b/trajectory/models/gpt/gpt.py#L32

While it is a small bug, I think it can have a great impact on performance and should be fixed. But it will break all pretrained checkpoints and I don't have time to retrain them for now.

### PR DESCRIPTION
**Pull request by Howuhh**
- [x] Fix bug
- [ ] Retrain models with default configs
- [ ] Check performance (should be same or better)
- [x] [Optional] Add citation bibtex to the README

### CODE REVIEW

**Comment by ezhang7423** [comment]
i am willing to retrain this in a few weeks after icml haha

**Comment by Howuhh** [comment]
Hi @ezhang7423! Very much appreciated, I am unfortunately also busy preparing for ICML and was doing other stuff for ICRL before that. 

In general, I think "faster" is a bit outdated here, simply adding tricks from [nanoGPT](https://github.com/karpathy/nanoGPT) + torch.compile + flash attention + torch.backends.cuda.matmul.allow_tf32 can speed up very significantly, maybe even faster than caching. 

Unfortunately I'm a bit disappointed in this path and prefer DT more now (which I also implemented in single-file format in [CORL](https://github.com/tinkoff-ai/CORL/blob/main/algorithms/dt.py), although without nanoGPT tricks yet)

**Comment by ezhang7423** [comment]
Hey thank you so much for the response. Can you comment on why you prefer DT? I am still making up my mind between the two. From my understanding, beam search can't be implemented in DT since it uses recent context rather than causal attention, so it requires recomputing the encodings of states and actions at every step. 

Can you also comment on the specific tricks from nanoGPT that are most useful?

**Comment by Howuhh** [comment]
Although I find beam search planning interesting, it is expensive, time-consuming (also hard to write it without bugs from my experience in other non-public and failed proposals), and has less interesting planning applications than diffusion alternatives (Decision Diffuser, Diffusion-QL, etc.).

My main concern is discretization and planning on all dimensions. Firstly, attention is already quite slow and quadratic and we inflate the length of the sequence even further from  N * 3 to N * (state_dim + action_dim + 1) which is huge. But, this it not only for learning on single context like in DT, but also for planning: context_len + planning_len!! On difficult tasks at least up to 3 steps ahead or higher is needed.  Caching can help here, but it is still not fast enough on harder tasks like robomimic (which I tried). Also, not sure how to use it on image-based tasks. Secondly, we can lose precision while decoding discretized tokens back and this can affect performance. For example, despite the fact that within the model we made the valid plan, during decoding such errors can compound causing additional distributional shift (this is also my hypothesis why TT not be able to do imitation learning on halfcheetah, see https://github.com/jannerm/trajectory-transformer/issues/10). 

> beam search can't be implemented in DT since it uses recent context rather than causal attention

No, beam search is just a general decoding algorithm which can be used for any next-token-prediction model, at least if prediction is a categorical distribution (no clue how to write it for continuous case). DT does not predict next tokens (only actions), and even when it does it is done with simply MSE loss. But both DT and TT are using causal attention. TT is just DT which uses discretization on states and action dimensions, then flattens it all within one sequence and predict next discrete token. After that planning can be used (or sampling). 

> it requires recomputing the encodings of states and actions at every step

Actually, attention caching can be implemented for DT too! It will be used during inference in the environment to speed up evaluation and avoid unnecessary computations on previous context (which is already been processed). 

Also, I'm a big fan of a simple but surprising DT extension: https://arxiv.org/abs/2210.14215

> Can you also comment on the specific tricks from nanoGPT that are most useful?

I think best way would be read original twitter thread from **karpathy**. He provides there some benchmarks for GPT training step time.

**Comment by ezhang7423** [comment]
> No, beam search is just a general decoding algorithm which can be used for any next-token-prediction model, at least if prediction is a categorical distribution (no clue how to write it for continuous case). 

To do this, we could simply predict the parameters of a multivariate gaussian or any other contrinuous distribution. Now that we have a probability distribution, we can sample from this to do beam search, assuming that we have a way to evaluate the probabilities of trajectories similar to evaluating probabilities of sentences in NLP.

> But both DT and TT are using causal attention. 

Are you sure? From my reading of the gym version of DT, it doesn't seem like they are doing causal attention [here](https://github.com/kzl/decision-transformer/blob/e2d82e68f330c00f763507b3b01d774740bee53f/gym/experiment.py#L154).  In fact, it seems like they will just pass the entire recent context in for each action prediction, which certainly seems wrong to me. I'm probably mistaken here, so if you can point me to where they are doing the causal attention I'd be extremely grateful. 

> Actually, attention caching can be implemented for DT too! It will be used during inference in the environment to speed up evaluation and avoid unnecessary computations on previous context (which is already been processed).

Is there an implementation of this that you can point me to? This should certainly be true if they're doing causal attention. The paper you referenced doesn't seem to directly mention caching.

> > Can you also comment on the specific tricks from nanoGPT that are most useful?

> I think best way would be read original twitter thread from karpathy.

Thanks. I've attached the relevant tweet here for future reference and other interested onlookers.

<blockquote class="twitter-tweet"><p lang="en" dir="ltr">having fun optimizing minGPT today<br>- base: 495ms<br>- zero_grad(set_to_none=True): 492<br>- torch.jit.script gelu: 463<br>- OMP_PROC_BIND=CLOSE: 453<br>- torch.backends.cuda.matmul.allow_tf32: 143<br>- torch.autocast(torch.bfloat16): 121<br>- FlashAttention: 102<br>now: more fused kernels more better</p>&mdash; Andrej Karpathy (@karpathy) <a href="https://twitter.com/karpathy/status/1607791537978748929?ref_src=twsrc%5Etfw">December 27, 2022</a>

**Comment by Howuhh** [comment]
@ezhang7423 I'm pretty sure of it. The piece of code to which your link points simply creates padding mask which is an addition to causal and they can stack together. For actual attention mask you need to look into Attention implementation:
1. gym version: https://github.com/kzl/decision-transformer/blob/e2d82e68f330c00f763507b3b01d774740bee53f/gym/decision_transformer/models/trajectory_gpt.py#L168
2. atari version: https://github.com/kzl/decision-transformer/blob/e2d82e68f330c00f763507b3b01d774740bee53f/atari/mingpt/model_atari.py#L75

Screenshot 2023-01-07 at 12 59 58

> Is there an implementation of this that you can point me to? 

No, but this is pretty straightforward to implement. Block will be identical to [TransformerBlock implementation](TransformerBlock) in this repo (without value masking on line 28 tho). 

Hope this helps!

**Comment by Howuhh** [comment]
resolved by https://github.com/Howuhh/faster-trajectory-transformer/pull/10
